File: src/main/java/io/gittul/infra/auth/jwt/JwtAuthenticator.java
package io.gittul.infra.auth.jwt;

import io.gittul.domain.user.UserInquiryService;
import io.gittul.domain.user.entity.User;
import io.gittul.infra.auth.exception.AuthenticationException;
import jakarta.servlet.http.HttpServletRequest;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

@Component
@RequiredArgsConstructor
public class JwtAuthenticator {
    private final JwtProvider jwtProvider;
    private final UserInquiryService userService; // Todo. ThreadLocal 로 변경 고려

    public User authenticate(HttpServletRequest request) {
        String authorizationHeader = request.getHeader("Authorization");
        String token = extractToken(authorizationHeader);

        TokenUserInfo tokenUserInfo = jwtProvider.validateToken(token);
        return userService.getUserFromTokenInfo(tokenUserInfo);
    }

    private String extractToken(String token) {
        if (token == null) throw AuthenticationException.TOKEN_NOT_FOUND;
        if (!token.startsWith("Bearer ")) throw AuthenticationException.INVALID_GRANT_TYPE;

        return token.substring(7);
    }
}

File: app/src/main/kotlin/io/gittul/app/infra/auth/jwt/JwtAuthenticator.kt
package io.gittul.app.infra.auth.jwt

import io.gittul.app.domain.user.UserInquiryService
import io.gittul.app.infra.auth.exception.AuthenticationException
import io.gittul.core.domain.user.entity.User
import jakarta.servlet.http.HttpServletRequest
import org.springframework.stereotype.Component

@Component
class JwtAuthenticator(
    private val jwtProvider: JwtProvider,
    private val userService: UserInquiryService,
    private val accessTokenProvider: AccessTokenProvider,
    private val refreshTokenProvider: RefreshTokenProvider
) {

    fun authenticate(request: HttpServletRequest): User {
        val token = extractToken(request) ?: throw AuthenticationException.INVALID_TOKEN
        return accessTokenProvider.validateToken(token)
    }

    fun extractRefreshToken(request: HttpServletRequest): String? {
        return request.cookies?.find { it.name == "refresh_token" }?.value
    }

    private fun extractToken(request: HttpServletRequest): String? {
        // 1. 쿠키에서 토큰 확인
        val cookies = request.cookies
        if (cookies != null) {
            for (cookie in cookies) {
                if (cookie.name == "access_token") {
                    return cookie.value
                }
            }
        }

        // 2. Authorization 헤더에서 토큰 확인
        val authHeader = request.getHeader("Authorization")
        if (authHeader != null && authHeader.startsWith("Bearer ")) {
            return authHeader.substring(7)
        }

        return null
    }
}
